### PR TITLE
[buteo-sync-plugin-caldav] Correctly detect added persistent exception occurrences. Contributes to MER#1195

### PR DESCRIPTION
**Pull request by chriadam**
This commit ensures that we detect locally added persistent exception
occurrences, and that we then upsync them as modifications (to the
original series).

Contributes to MER#1195

### CODE REVIEW

**Comment by chriadam** [comment]
@pvuorela @blammit

File: src/notebooksyncagent.cpp [hunk 1]
```diff
@@ -761,6 +762,16 @@ bool NotebookSyncAgent::calculateDelta(
     // separate them into buckets.
     // note that each remote URI can be associated with multiple local incidences (due recurrenceId incidences)
     // Here we can determine local additions and remote deletions.
+    KCalCore::Incidence::List additions, additionsSyncDate, addedPersistentExceptionOccurrences;
+    if (!mStorage->insertedIncidences(&additions, fromDate, mNotebook->uid()) ||
+        !mStorage->insertedIncidences(&additionsSyncDate, syncDateTime, mNotebook->uid())) {
+        LOG_CRITICAL("mKCal::ExtendedStorage::insertedIncidences() failed");
+        return false;
+    }
+    uniteIncidenceLists(additionsSyncDate, &additions);
```
Comment (pvuorela) [inline src/notebooksyncagent.cpp]: Should these lines be equal to just fetching inserted incidences with earlier time of the two?

**Comment by pvuorela** [comment]
LGTM

File: src/notebooksyncagent.cpp [hunk 2]
```diff
@@ -804,31 +836,52 @@ bool NotebookSyncAgent::calculateDelta(
...
-            if (uriWasEmpty) {
-                // we originally upsynced this pure-local addition, but then connectivity was
-                // lost before we updated the uid of it locally to include the remote uri.
-                // subsequently, the user deleted the incidence.
-                // Hence, it exists remotely, and has been deleted locally.
-                LOG_DEBUG("have local deletion for partially synced incidence:" << incidence->uid() << incidence->recurrenceId().toString());
+    // i = 1, deal with deletions of entire series / non-recurrence-id incidences
+    // i = 2, deal with deletions of recurrence-id incidences (persistent exception occurrences)
+    QSet<QString> deletedSeriesUids;
+    for (int i = 1; i <= 2; ++i) {
```
Comment (blammit) [inline src/notebooksyncagent.cpp]: Seems odd to run this in a loop, could just call a function twice instead?

File: src/notebooksyncagent.cpp (hunk 1 above)
Comment (chriadam) [inline src/notebooksyncagent.cpp]: You're absolutely right.  I'll fix this (and the other cases where I did similar).

File: src/notebooksyncagent.cpp (hunk 2 above)
Comment (chriadam) [inline src/notebooksyncagent.cpp]: Fair enough.  I can macroise it if that'd be better.  Splitting it into a function seems suboptimal as it's not really any more "special" than any other codeblock in this function, but I can see why the loop is confusing.

Comment (blammit) [inline src/notebooksyncagent.cpp]: Personally I find functions easier to read than macros once it gets beyond a few lines, and also shortening a method by separating a large block into a private function makes the code easier to follow, but ymmv.

File: src/notebooksyncagent.cpp [hunk 3]
```diff
@@ -750,6 +751,7 @@ bool NotebookSyncAgent::calculateDelta(
     // and createdBefore; so instead we have to build a datetime which "should" satisfy
     // the inequality for all possible local modifications detectable since the last sync.
     KDateTime syncDateTime = mNotebook->syncDate().addSecs(1); // deleted after, created before...
+    KDateTime earliestAnchorTimestamp = syncDateTime < fromDate ? syncDateTime : fromDate;
```
Comment (chriadam) [inline src/notebooksyncagent.cpp]: Hrm... Actually... I'm not sure about this after all...

mkcal uses an inequality: SELECT * from Components WHERE created < fromDate AND modified > fromDate;

Similarly for deletions: WHERE created < fromDate AND deleted > fromDate;

It's possible that we NEED to do the unite in order to catch both the ones which were created before the earlier of the two timestamps, AS WELL as the ones which were created AFTER the earlier timestamp but BEFORE the later timestamp...

Comment (pvuorela) [inline src/notebooksyncagent.cpp]: Yea, could be for the later cases, but first insertedIncidences should be fine?

File: src/notebooksyncagent.cpp [hunk 4]
```diff
@@ -782,60 +792,100 @@ bool NotebookSyncAgent::calculateDelta(
...
-            }
-            LocalDeletion localDeletion(incidence, remoteUriEtags.value(remoteUri), remoteUri);
-            localDeletions->append(localDeletion);
-            seenRemoteUris.insert(remoteUri);
-        } else {
-            // it was either already deleted remotely, or was never upsynced from the local prior to deletion.
-            LOG_DEBUG("ignoring local deletion of non-existent remote incidence:" << incidence->uid() << incidence->recurrenceId().toString() << "at" << remoteUri);
-        }
+    QSet<QString> deletedSeriesUids;
+    static const int BASE_SERIES = 1, PERSISTENT_EXCEPTION_OCCURRENCES = 0;
```
Comment (pvuorela) [inline src/notebooksyncagent.cpp]: As personal preference, I'd consider a for loop as lesser evil than near 40 line macro inside a function.

File: src/notebooksyncagent.cpp (hunk 3 above)
Comment (chriadam) [inline src/notebooksyncagent.cpp]: You're right.  Will undo the changes to the other ones, and use the earlier timestamp for the call to insertedIncidences here.  Thanks.

File: src/notebooksyncagent.cpp (hunk 4 above)
Comment (chriadam) [inline src/notebooksyncagent.cpp]: I've refactored this, as I found another case which could cause problems.  Now I only loop over them once here, but later iterate over the set and erase ones which need erasing, AFTER the set of remote modifications is also determined.  On the plus side, the code is nicer too.

**Comment by blammit** [comment]
LGTM
